scipy/signal/signaltools.py

```diff
@@ -1730,23 +1729,65 @@ def resample(x, num, t=None, axis=0, window=None):
             W = window
         else:
             W = fftpack.ifftshift(get_window(window, Nx))
-        newshape = [1] * x.ndim
-        newshape[axis] = len(W)
-        W.shape = newshape
-        X = X * W
+
+        newshape_W = [1] * x.ndim
+        newshape_W[axis] = len(W)
+
     sl = [slice(None)] * len(x.shape)
     newshape = list(x.shape)
     newshape[axis] = num
-    N = int(np.minimum(num, Nx))
-    Y = zeros(newshape, 'D')
-    sl[axis] = slice(0, (N + 1) // 2)
-    Y[sl] = X[sl]
-    sl[axis] = slice(-(N - 1) // 2, None)
-    Y[sl] = X[sl]
-    y = fftpack.ifft(Y, axis=axis) * (float(num) / float(Nx))
-
-    if x.dtype.char not in ['F', 'D']:
-        y = y.real
+
+    # Can we use faster real FFT?
+    if not np.issubdtype(x.dtype, complex) and (_rfft_mt_safe or _rfft_lock.acquire(False)):
+        try:
+            X = fftpack.rfft(x, axis=axis)
+            if window is not None:
+                # Reorder the window to fit the order rfft produces its output
+                W_real = W.copy()
+                for i in range(1, Nx//2):
+                    W_real[2*i] = W[i]
+                    W_real[2*i-1] = W[i]
+                # Deal with the differnet length for odd/even Nx
+                W_real[Nx-1:] = W[Nx//2]
+
+                X = X * W_real.reshape(newshape_W)
+            Y = zeros(newshape)
+
+            if num > Nx:
+                if Nx % 2:
+                    sl[axis] = slice(0, Nx)
+                    Y[sl] = X[sl]
+                else:
+                    # X[-1] is only the real part
+                    sl[axis] = slice(0, Nx-1)
+                    Y[sl] = X[sl]
+
+                    sl = [slice(None)] * len(x.shape)
+                    sl[axis] = slice(Nx-1, Nx)
+                    Y[sl] = 0.5*X[sl]
+            else:
+                sl[axis] = slice(0, num)
+                Y[sl] = X[sl]     
+
+            y = fftpack.irfft(Y, axis=axis) * (float(num) / float(Nx))
+
+        finally:
+            if not _rfft_mt_safe:
+                _rfft_lock.release()
+
+    # Full complex FFT
+    else:
+        X = fftpack.fft(x, axis=axis)
+        X = X if window is None else X * W.reshape(newshape_W)
```

Comment:
while we're digging around in here, we can avoid an effective memcopy of `X` by changing this line to:
```
if window is not None:
    X *= W.reshape(newshape_W)
```